File: sudoku.rb
require 'pry'

class Sudoku

    attr_accessor :solution, :board

    def initialize( number_of_holes )
        self.solution = [ nil ]
        self.solution = populate_puzzle( starting_puzzle ) until !self.solution.flatten.include?( nil )
        self.board = poke( self.solution, generate_grille( 48 ) )
    end

    def self.to_s( board )
        %(
┏━━━┳━━━┳━━━┱───┬───┬───┲━━━┳━━━┳━━━┓
┃ #{ board[ 0 ][ 0 ].zero? ? " " : board[ 0 ][ 0 ] } ┃ #{ board[ 0 ][ 1 ].zero? ? " " : board[ 0 ][ 1 ] } ┃ #{ board[ 0 ][ 2 ].zero? ? " " : board[ 0 ][ 2 ] } ┃ #{ board[ 0 ][ 3 ].zero? ? " " : board[ 0 ][ 3 ] } │ #{ board[ 0 ][ 4 ].zero? ? " " : board[ 0 ][ 4 ] } │ #{ board[ 0 ][ 5 ].zero? ? " " : board[ 0 ][ 5 ] } ┃ #{ board[ 0 ][ 6 ].zero? ? " " : board[ 0 ][ 6 ] } ┃ #{ board[ 0 ][ 7 ].zero? ? " " : board[ 0 ][ 7 ] } ┃ #{ board[ 0 ][ 8 ].zero? ? " " : board[ 0 ][ 8 ] } ┃
┣━━━╋━━━╋━━━╉───┼───┼───╊━━━╋━━━╋━━━┫
┃ #{ board[ 1 ][ 0 ].zero? ? " " : board[ 1 ][ 0 ] } ┃ #{ board[ 1 ][ 1 ].zero? ? " " : board[ 1 ][ 1 ] } ┃ #{ board[ 1 ][ 2 ].zero? ? " " : board[ 1 ][ 2 ] } ┃ #{ board[ 1 ][ 3 ].zero? ? " " : board[ 1 ][ 3 ] } │ #{ board[ 1 ][ 4 ].zero? ? " " : board[ 1 ][ 4 ] } │ #{ board[ 1 ][ 5 ].zero? ? " " : board[ 1 ][ 5 ] } ┃ #{ board[ 1 ][ 6 ].zero? ? " " : board[ 1 ][ 6 ] } ┃ #{ board[ 1 ][ 7 ].zero? ? " " : board[ 1 ][ 7 ] } ┃ #{ board[ 1 ][ 8 ].zero? ? " " : board[ 1 ][ 8 ] } ┃
┣━━━╋━━━╋━━━╉───┼───┼───╊━━━╋━━━╋━━━┫
┃ #{ board[ 2 ][ 0 ].zero? ? " " : board[ 2 ][ 0 ] } ┃ #{ board[ 2 ][ 1 ].zero? ? " " : board[ 2 ][ 1 ] } ┃ #{ board[ 2 ][ 2 ].zero? ? " " : board[ 2 ][ 2 ] } ┃ #{ board[ 2 ][ 3 ].zero? ? " " : board[ 2 ][ 3 ] } │ #{ board[ 2 ][ 4 ].zero? ? " " : board[ 2 ][ 4 ] } │ #{ board[ 2 ][ 5 ].zero? ? " " : board[ 2 ][ 5 ] } ┃ #{ board[ 2 ][ 6 ].zero? ? " " : board[ 2 ][ 6 ] } ┃ #{ board[ 2 ][ 7 ].zero? ? " " : board[ 2 ][ 7 ] } ┃ #{ board[ 2 ][ 8 ].zero? ? " " : board[ 2 ][ 8 ] } ┃
┡━━━╇━━━╇━━━╋━━━╈━━━╈━━━╋━━━╇━━━╇━━━┩
│ #{ board[ 3 ][ 0 ].zero? ? " " : board[ 3 ][ 0 ] } │ #{ board[ 3 ][ 1 ].zero? ? " " : board[ 3 ][ 1 ] } │ #{ board[ 3 ][ 2 ].zero? ? " " : board[ 3 ][ 2 ] } ┃ #{ board[ 3 ][ 3 ].zero? ? " " : board[ 3 ][ 3 ] } ┃ #{ board[ 3 ][ 4 ].zero? ? " " : board[ 3 ][ 4 ] } ┃ #{ board[ 3 ][ 5 ].zero? ? " " : board[ 3 ][ 5 ] } ┃ #{ board[ 3 ][ 6 ].zero? ? " " : board[ 3 ][ 6 ] } │ #{ board[ 3 ][ 7 ].zero? ? " " : board[ 3 ][ 7 ] } │ #{ board[ 3 ][ 8 ].zero? ? " " : board[ 3 ][ 8 ] } │
├───┼───┼───╊━━━╋━━━╋━━━╉───┼───┼───┤
│ #{ board[ 4 ][ 0 ].zero? ? " " : board[ 4 ][ 0 ] } │ #{ board[ 4 ][ 1 ].zero? ? " " : board[ 4 ][ 1 ] } │ #{ board[ 4 ][ 2 ].zero? ? " " : board[ 4 ][ 2 ] } ┃ #{ board[ 4 ][ 3 ].zero? ? " " : board[ 4 ][ 3 ] } ┃ #{ board[ 4 ][ 4 ].zero? ? " " : board[ 4 ][ 4 ] } ┃ #{ board[ 4 ][ 5 ].zero? ? " " : board[ 4 ][ 5 ] } ┃ #{ board[ 4 ][ 6 ].zero? ? " " : board[ 4 ][ 6 ] } │ #{ board[ 4 ][ 7 ].zero? ? " " : board[ 4 ][ 7 ] } │ #{ board[ 4 ][ 8 ].zero? ? " " : board[ 4 ][ 8 ] } │
├───┼───┼───╊━━━╋━━━╋━━━╉───┼───┼───┤
│ #{ board[ 5 ][ 0 ].zero? ? " " : board[ 5 ][ 0 ] } │ #{ board[ 5 ][ 1 ].zero? ? " " : board[ 5 ][ 1 ] } │ #{ board[ 5 ][ 2 ].zero? ? " " : board[ 5 ][ 2 ] } ┃ #{ board[ 5 ][ 3 ].zero? ? " " : board[ 5 ][ 3 ] } ┃ #{ board[ 5 ][ 4 ].zero? ? " " : board[ 5 ][ 4 ] } ┃ #{ board[ 5 ][ 5 ].zero? ? " " : board[ 5 ][ 5 ] } ┃ #{ board[ 5 ][ 6 ].zero? ? " " : board[ 5 ][ 6 ] } │ #{ board[ 5 ][ 7 ].zero? ? " " : board[ 5 ][ 7 ] } │ #{ board[ 5 ][ 8 ].zero? ? " " : board[ 5 ][ 8 ] } │
┢━━━╈━━━╈━━━╋━━━╇━━━╇━━━╋━━━╈━━━╈━━━┪
┃ #{ board[ 6 ][ 0 ].zero? ? " " : board[ 6 ][ 0 ] } ┃ #{ board[ 6 ][ 1 ].zero? ? " " : board[ 6 ][ 1 ] } ┃ #{ board[ 6 ][ 2 ].zero? ? " " : board[ 6 ][ 2 ] } ┃ #{ board[ 6 ][ 3 ].zero? ? " " : board[ 6 ][ 3 ] } │ #{ board[ 6 ][ 4 ].zero? ? " " : board[ 6 ][ 4 ] } │ #{ board[ 6 ][ 5 ].zero? ? " " : board[ 6 ][ 5 ] } ┃ #{ board[ 6 ][ 6 ].zero? ? " " : board[ 6 ][ 6 ] } ┃ #{ board[ 6 ][ 7 ].zero? ? " " : board[ 6 ][ 7 ] } ┃ #{ board[ 6 ][ 8 ].zero? ? " " : board[ 6 ][ 8 ] } ┃
┣━━━╋━━━╋━━━╉───┼───┼───╊━━━╋━━━╋━━━┫
┃ #{ board[ 7 ][ 0 ].zero? ? " " : board[ 7 ][ 0 ] } ┃ #{ board[ 7 ][ 1 ].zero? ? " " : board[ 7 ][ 1 ] } ┃ #{ board[ 7 ][ 2 ].zero? ? " " : board[ 7 ][ 2 ] } ┃ #{ board[ 7 ][ 3 ].zero? ? " " : board[ 7 ][ 3 ] } │ #{ board[ 7 ][ 4 ].zero? ? " " : board[ 7 ][ 4 ] } │ #{ board[ 7 ][ 5 ].zero? ? " " : board[ 7 ][ 5 ] } ┃ #{ board[ 7 ][ 6 ].zero? ? " " : board[ 7 ][ 6 ] } ┃ #{ board[ 7 ][ 7 ].zero? ? " " : board[ 7 ][ 7 ] } ┃ #{ board[ 7 ][ 8 ].zero? ? " " : board[ 7 ][ 8 ] } ┃
┣━━━╋━━━╋━━━╉───┼───┼───╊━━━╋━━━╋━━━┫
┃ #{ board[ 8 ][ 0 ].zero? ? " " : board[ 8 ][ 0 ] } ┃ #{ board[ 8 ][ 1 ].zero? ? " " : board[ 8 ][ 1 ] } ┃ #{ board[ 8 ][ 2 ].zero? ? " " : board[ 8 ][ 2 ] } ┃ #{ board[ 8 ][ 3 ].zero? ? " " : board[ 8 ][ 3 ] } │ #{ board[ 8 ][ 4 ].zero? ? " " : board[ 8 ][ 4 ] } │ #{ board[ 8 ][ 5 ].zero? ? " " : board[ 8 ][ 5 ] } ┃ #{ board[ 8 ][ 6 ].zero? ? " " : board[ 8 ][ 6 ] } ┃ #{ board[ 8 ][ 7 ].zero? ? " " : board[ 8 ][ 7 ] } ┃ #{ board[ 8 ][ 8 ].zero? ? " " : board[ 8 ][ 8 ] } ┃
┗━━━┻━━━┻━━━┹───┴───┴───┺━━━┻━━━┻━━━┛
        )
    end

    private

    def populated_region; ( 1..9 ).to_a.shuffle.each_slice( 3 ).to_a; end
    
    def empty_region; [ [ 0, 0, 0 ], [ 0, 0, 0 ], [ 0, 0, 0 ] ]; end
    
    def starting_puzzle
        [ populated_region.zip( empty_region, empty_region ), empty_region.zip( populated_region, empty_region ), empty_region.zip( empty_region, populated_region ) ].flatten.each_slice( 9 ).to_a
    end

    def populate_puzzle( puzzle )
        for row in ( 0..8 ) do
            for column in ( 0..8 ) do
                if puzzle[ row ][ column ].zero?
                    this_row = puzzle[ row ]
                    this_column = puzzle.map{ | row | row[ column ] }
                    this_region = [ puzzle[ row / 3 * 3 ].slice( column / 3 * 3, 3 ), puzzle[ ( row / 3 * 3 ) + 1 ].slice( column / 3 * 3, 3 ), puzzle[ ( row / 3 * 3 ) + 2 ].slice( column / 3 * 3, 3 ) ].flatten
                    puzzle[ row ][ column ] = ( 1..9 ).to_a.reject{ | digit | this_row.union( this_column, this_region ).include?( digit ) }.sample
                end
            end
        end
        puzzle
    end
    
    def generate_grille( number_of_holes )
        ( Array.new( 81 - number_of_holes, false ) + Array.new( number_of_holes, true ) ).shuffle
    end
    
    def poke( puzzle, grille )
        return puzzle.flatten.map.with_index{ | cell, index | grille[ index ] ? 0 : cell }.each_slice( 9 ).to_a
    end

end

test = Sudoku.new( 64 )

binding.pry
false

File: sudoku.js
/*
function shuffle( array ) {
    for ( let i = array.length - 1; i > 0; i-- ) {
        const j = Math.floor( Math.random() * ( i + 1 ) );
        [ array[ i ], array[ j ] ] = [ array[ j ], array[ i ] ];
    }
    return array;
}

function populatedRegion() {
    const shuffleDigits = shuffle( [ 1, 2, 3, 4, 5, 6, 7, 8, 9 ] );
    return [ shuffleDigits.slice( 0, 3 ), shuffleDigits.slice( 3, 6 ), shuffleDigits.slice( 6 ) ];
}

function emptyRegion() {
    return [ [ 0, 0, 0 ], [ 0, 0, 0 ], [ 0, 0, 0 ] ];
}

function startingPuzzle() {
    const startingQuadrant1 = populatedRegion(), startingQuadrant2 = populatedRegion(), startingQuadrant3 = populatedRegion();
    const firstThreeRows =  startingQuadrant1.map( ( row, index ) => { return row.concat( emptyRegion()[ index ] ).concat( emptyRegion()[ index ] ) } );
    const secondThreeRows =  emptyRegion().map( ( row, index ) => { return row.concat( startingQuadrant2[ index ] ).concat( emptyRegion()[ index ] ) } );
    const thirdThreeRows =  emptyRegion().map( ( row, index ) => { return row.concat( emptyRegion()[ index ] ).concat( startingQuadrant3[ index ] ) } );
    return firstThreeRows.concat( secondThreeRows ).concat( thirdThreeRows );
}

function populate( puzzle ) {
    for ( const row of [ 0, 1, 2, 3, 4, 5, 6, 7, 8 ] ) {
        for ( const column of [ 0, 1, 2, 3, 4, 5, 6, 7, 8 ] ) {
            if ( !puzzle[ row ][ column ] ) {
                let thisRow = puzzle[ row ];
                let thisColumn = puzzle.map( row => row[ column ] );
                let thisRegion = puzzle[ Math.floor( row / 3 ) * 3 ].slice( Math.floor( column / 3 ) * 3, ( Math.floor( column / 3 ) * 3 ) + 3 );
                thisRegion = thisRegion.concat( puzzle[ ( Math.floor( row / 3 ) * 3 ) + 1 ].slice( Math.floor( column / 3 ) * 3, ( Math.floor( column / 3 ) * 3 ) + 3 ) );
                thisRegion = thisRegion.concat( puzzle[ ( Math.floor( row / 3 ) * 3 ) + 2 ].slice( Math.floor( column / 3 ) * 3, ( Math.floor( column / 3 ) * 3 ) + 3 ) );
                let alreadyChosenDigits = [ ...new Set( thisRow.concat( thisColumn ).concat( thisRegion ) ) ].filter( digit => !!digit );
                let possibleDigits = [ 1, 2, 3, 4, 5, 6, 7, 8, 9 ].filter( digit => !alreadyChosenDigits.includes( digit ) );
                puzzle[ row ][ column ] = possibleDigits[ Math.floor( Math.random() * possibleDigits.length ) ];
            }
        }
    }
}

function grille( numberOfHoles ) {
    return shuffle( Array( numberOfHoles ).fill( true ).concat( Array( 81 - numberOfHoles ).fill( false ) ) );
}

function poke( puzzle, grille ) {
    const flattenedAndPoked = puzzle.flat().map( ( digit, index ) => grille[ index ] ? 0 : digit );
    result = [];
    for ( let i = 0; i < 81; i += 9 ) { result.push( flattenedAndPoked.slice( i, i + 9 ) ) }
    return result;
}

//////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////
//////////////////////////////////////////////////////////////////////////////////////////////////////

let solution = [ undefined ];
while ( solution.flat().includes( undefined ) ) {
    solution = startingPuzzle();
    populate( solution );
}
let board = poke( solution, grille( 32 ) );
for ( const row of [ 0, 1, 2, 3, 4, 5, 6, 7, 8 ] ) {
    for ( const column of [ 0, 1, 2, 3, 4, 5, 6, 7, 8 ] ) {
        if ( !!board[ row ][ column ] ) {
            allCells[ row ][ column ].firstElementChild.classList.add( "clue" );
            allCells[ row ][ column ].firstElementChild.disabled = true;
            allCells[ row ][ column ].firstElementChild.value = board[ row ][ column ]
        }
    }
}
*/

const allCells = [
    [ ...document.querySelector( '[data-row="0"]' ).getElementsByTagName( "td" ) ],
    [ ...document.querySelector( '[data-row="1"]' ).getElementsByTagName( "td" ) ],
    [ ...document.querySelector( '[data-row="2"]' ).getElementsByTagName( "td" ) ],
    [ ...document.querySelector( '[data-row="3"]' ).getElementsByTagName( "td" ) ],
    [ ...document.querySelector( '[data-row="4"]' ).getElementsByTagName( "td" ) ],
    [ ...document.querySelector( '[data-row="5"]' ).getElementsByTagName( "td" ) ],
    [ ...document.querySelector( '[data-row="6"]' ).getElementsByTagName( "td" ) ],
    [ ...document.querySelector( '[data-row="7"]' ).getElementsByTagName( "td" ) ],
    [ ...document.querySelector( '[data-row="8"]' ).getElementsByTagName( "td" ) ],
]

function peers( puzzle, row, column) {
    let result = puzzle[ row ];
    const topLeftCorner = [ Math.floor( row / 3 ) * 3, Math.floor( column / 3 ) * 3];
    result = result.concat( puzzle.map( row => row[ column ] ) );
    result = result.concat( [ puzzle[ topLeftCorner[ 0 ] ][ topLeftCorner[ 1 ] ],
        puzzle[ topLeftCorner[ 0 ] ][ topLeftCorner[ 1 ] + 1 ],
        puzzle[ topLeftCorner[ 0 ] ][ topLeftCorner[ 1 ] + 2 ],
        puzzle[ topLeftCorner[ 0 ] + 1 ][ topLeftCorner[ 1 ] ],
        puzzle[ topLeftCorner[ 0 ] + 1 ][ topLeftCorner[ 1 ] + 1 ],
        puzzle[ topLeftCorner[ 0 ] + 1 ][ topLeftCorner[ 1 ] + 2 ],
        puzzle[ topLeftCorner[ 0 ] + 2 ][ topLeftCorner[ 1 ] ],
        puzzle[ topLeftCorner[ 0 ] + 2 ][ topLeftCorner[ 1 ] + 1 ],
        puzzle[ topLeftCorner[ 0 ] + 2 ][ topLeftCorner[ 1 ] + 2 ] ] );
    return [ ...new Set( result ) ];
}

function clearHighlight() {
    allCells.flat().forEach( cell => cell.classList.remove( "highlight" ) );
}

function handleDomClick( documentClick ) {
    if ( documentClick.target.tagName != "INPUT" ) { clearHighlight(); }
}

function highlightCellPeers( cellClick ) {
    if ( cellClick.target.tagName === "INPUT" ) {
        clearHighlight();
        peers( allCells, parseInt( cellClick.target.closest( "tr" ).dataset.row ), parseInt( cellClick.target.dataset.column ) ).forEach( peer => peer.classList.add( "highlight" ) );
    }
}

document.addEventListener( "DOMContentLoaded", () => {
    document.addEventListener( "click", handleDomClick );
    document.getElementById( "sudoku-board" ).addEventListener( "click", highlightCellPeers );
} );
